decisions/migrations.md

```diff
@@ -0,0 +1,262 @@
+4. Record migration results
+5. Easy way to explore full schema (all of the migrations applied, for a particular version)
+6. Test data migrations
+7. Should work with RC versions
+8. (maybe) Consolidate common sql for PG/Sqlite (we have quite a few duplicate sql statements)
+9. (maybe) Use strongly typed diesel db if possible, for data migrations
+10. Migration code needs to run in isolation (no production code should run in parallel)
+
+## Examples/Extra
+
+### 2.a 
+
+There might be a need to do multiple schema and data updates within one version upgrade. For example if a required field is added that needs to be populated with some value derived from existing database values, we would want to:
+* add schema without the constraint
+* add the values 
+* add the constraint
+
+Another example is when sync buffer was consolidated:
+* Add new sync buffer
+* Copy rows from remote and central to the new sync buffer (see proposed solution, that touches on `9.`)
+* Remove old sync buffer
+
+Above can also be done by keeping existing sync buffer in memory than doing one schema migration with adding and removing sync buffer then populating new sync buffer from memory. 
+
+I think it's better to keep migrations flexible, without a particular pattern of operation sequence.
+
+### 2.b
+
+If we add a new field/table in central and remote is not up to date, the field/table will still be synced and recorded in sync buffer. And when we update to a new version of remote that has this new field/table we can just crawl through sync buffer and try to re-integrate this new field/table. This avoids needing to re-sync this info, and we can avoid implementing this mechanism
+
+Remote data should be safe with above example, as central shouldn't be updating foreign data.
+
+### 4.
+
+General data about migration result should be logged for debugging purposes, and on migration failure we can't guarantee correct behaviour of the app, so we would need to either disable server and not allow it to start again or revert to previous data version and inform the admin/user that app can be downgraded. 
+
+I think ability to revert to previous version (on failure) allows for system to be available to the users in timely manner, in case errors in migration occur.
+
+### 6.
+
+This could be tricky especially if we want to use diesel db strong typing. I think this in necessary and very important, we've had some major failures in the past with data migrations.
+
+### 7.
+
+Say we are on version 1.01, and are working on version 1.02, before deployment we should be able to:
+* Make an RC version, say 1.02-RC1
+* Test upgrade of 1.01 data file to 1.02-RC1 (just by replacing binary)
+* RC versions should not be upgradable, if you open RC version and there is a version mismatch then user should be warned that version are incompatible (this is to avoid additions to migrations that won't run or won't be tested)
+* When we happy with 1.02, it's released and 1.01 would be upgradable to 1.02-RC1
```

Comment:
I'm not understanding this section will probably need you to explain it to me.. 
as an attempt to explain myself.. my confusion is:
- upgrading a data file (should that be database?) by replacing binary (is that the server binary?) Are you saying that, in this example, 1.01 can be upgraded to 1.02-RC1 but then the RC version isn't upgradeable? You'd have to go back to 1.01 to try out 1.02 when it is released? Or to test 1.02-RC2?
- wasn't sure when a version mismatch error would be shown. From reading below, it's when the executable version is lower than the database version. Is that right?
- the last bullet point: why would you upgrade to 1.02-RC1 when 1.02 has been released?

Sorry! this is me not understanding - I expect that it all makes sense and is probably what I have in mind too

Reply:
Yeah this is a bit tricky, hmm, i've tried to explain further: 8cdd9af87103132e203835e4d8f800e370588f2c

My worry is that quite a few uses cases arise if we allow RC versions to be further migrated (to another RC version or higher non RC version). 

> the last bullet point: why would you upgrade to 1.02-RC1 when 1.02 has been released?

Oops that should have been 1.02 (not 1.02-RC1).

> upgrading a data file (should that be database?)

🙏 changed to database in the latest commit

> Are you saying that, in this example, 1.01 can be upgraded to 1.02-RC1 but then the RC version isn't upgradeable?

Yeah

> You'd have to go back to 1.01 to try out 1.02 when it is released? Or to test 1.02-RC2?

Yep, when 1.02-RC2 is made, you will have to test it by opening 1.01 (can't try it again a database that's on 1.02-RC1).

> Sorry! this is me not understanding - I expect that it all makes sense and is probably what I have in mind too

I think I over complicated it with my example, 🤞 it makes sense now